---
title: "RNA_setup"
author: "Lorenzo.MCM"
date: "2024-06-04"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Initial Setup

## Installing and loading packages.

```{r eval=FALSE}

install.packages(c('dplyr','biomaRt','readxl','DESeq2'), repos = "http://cran.us.r-project.org" )
```

```{r eval=FALSE}
BiocManager::install(c('DESeq2', 'AnnotationDbi', 'org.Mm.eg.db', 'apeglm', 'ReportingTools', 'clusterProfiler'))
```

```{r message=FALSE}
library(readxl)
library(DESeq2)
library(vsn)
library(Glimma)
library(regionReport)
library(clusterProfiler)
library(org.Mm.eg.db)
library(AnnotationDbi)
library(biomaRt)
library(stringr)
library(dplyr)
```

## Importing data and tidying up resulting data frame so rows are ID'd by Gene Symbol.NAs omitted for comparison with Gene list

```{r message=FALSE, warning=FALSE}

# Setting Working Directory 

#setwd("your/file/path")
setwd("C:/Users/Lorenzo/Documents/MESTRADO/R/RNAbps")

# Importing main reads file

data = as.data.frame(read_excel("Reads_genes.xlsx"))

# Keeping only the columns with MGI, RefSeq ID and Gene symbols + raw counts, omitting NAs 

reads = data %>% select(MGI, Gene_ID, Gene_Symbol, contains("Count"))

reads = as.data.frame(reads)

reads = na.omit(reads)

```

```{r}

# Importing MSGP database excel

MSGP = as.data.frame(read_excel("MSGP.xlsx"))

# Removing blank line and only keeping a list of entrez IDs

MSGP = MSGP[-1,]

MSGP = MSGP$`Entrez Gene ID`

MSGP = unique(MSGP)

# Importing ensembl database and listing relevant criteria for homologue df

ensembl = useMart("ensembl", dataset = "hsapiens_gene_ensembl")

human = useDataset("hsapiens_gene_ensembl", mart = ensembl)
mouse = useDataset("mmusculus_gene_ensembl", mart = ensembl)

hAttributes = c("ensembl_gene_id", "external_gene_name",
                "mmusculus_homolog_ensembl_gene", 
                "mmusculus_homolog_associated_gene_name")

# Creating homologue df out of ensembl filtered for ENS ID and name

orth.mouse = getBM(hAttributes, filters = "with_mmusculus_homolog", values = TRUE,
                   mart = human, uniqueRows = TRUE)

# Merging entrez and MGI IDs into df

mapping = getBM(attributes = c('ensembl_gene_id', 'entrezgene_id'),
                 filters = 'ensembl_gene_id',
                 values = orth.mouse$ensembl_gene_id,
                 mart = human)

orth.mouse = merge(orth.mouse, mapping, by.x = 'ensembl_gene_id', by.y = 'ensembl_gene_id', all.x = TRUE)

mapping_mmus = getBM(attributes = c("ensembl_gene_id", "entrezgene_id", "mgi_id"),
                     filters = "ensembl_gene_id",
                     values = orth.mouse$mmusculus_homolog_ensembl_gene,
                     mart = mouse)

mapping_mmus = mapping_mmus %>% rename(mmusculus_homolog_ensembl_gene = ensembl_gene_id)
mapping_mmus = mapping_mmus %>% rename(mmusculus_homolog_entrez_id = entrezgene_id)

orth.mouse = merge(orth.mouse, mapping_mmus, by.x = "mmusculus_homolog_ensembl_gene", by.y = "mmusculus_homolog_ensembl_gene", all.x = TRUE)

# Removing ENS IDs and MGI: prefix in all MGI entries

orth.mouse = orth.mouse[,-c(1,2)]

orth.mouse$mgi_id = gsub("MGI:", "", orth.mouse$mgi_id)

# Keeping only the genes whose entrez IDs are present in the MSGP entrez ID list

MSGP_ID = orth.mouse[orth.mouse$entrezgene_id %in% MSGP,]

MSGP_ID = unique(MSGP_ID)

# If analyzing human SG genes, uncomment this

#rbps = reads[reads$MGI %in% MSGP_ID$mgi_id,]

```

## Importing panther-db list of RBPs and tidying it up (browsed for RNA Metabolism Proteins and filtered for Mus musculus, exported to and edited in excel so the only row was MGI IDs, removed genes that did not have MGI IDs)

```{r message=FALSE, warning=FALSE}

# Importing list of RBPs, omitting NAs

#mainPanther = as.data.frame(read_excel('rna_met.xlsx', col_names = FALSE))

mainPanther = as.data.frame(read_excel('Panther-RBPs.xlsx', col_names = FALSE))

mainPanther = mainPanther[!is.na(mainPanther)]

# Filtering the reads object so it only contains genes whose MGI is also present in the list of RBPs, resulting in a final "rbps" object 
#(alternatively comment this first line and uncomment the last one to analyze all genes)

#comment this for analyzing all genes (uncomment last line)
rbps = reads[reads$MGI %in% mainPanther,]

#comment this for analyzing only rbps (uncomment previous line) 
#rbps = reads

```

## Naming conditions and final set up of rbps object

```{r message=FALSE, warning=FALSE}

# Naming rows in the rbps object after their MGI ID and removing all columns except those with the counts for each condition

rownames(rbps) = rbps$MGI

rbps = rbps[,-c(1,2,3)]

# Naming raw counts columns after their conditions, based on an external file

colnames(rbps) = (read_excel("conditions.xlsx", col_names = FALSE))[1,]
              
```

# Setting up for DESeq 
## Intermediate objects for selecting conditions for comparisons through DESeq

```{r message=FALSE, warning=FALSE}

#Splitting conditions for line and sex
rbpsF_WT = rbps %>% select(contains("FWT"))
rbpsM_WT = rbps %>% select(contains("MWT"))
rbpsF_TG = rbps %>% select(contains("F36"), contains("F41"))
rbpsM_TG = rbps %>% select(contains("M36"), contains("M41"))
rbpsF_36 = rbps %>% select(contains("F36"))
rbpsM_36 = rbps %>% select(contains("M36"))
rbpsF_41 = rbps %>% select(contains("F41"))
rbpsM_41 = rbps %>% select(contains("M41"))

#Final split into timepoints
T1F_41 = rbpsF_41[,grep("^1",names(rbpsF_41), value = TRUE)]
T1M_41 = rbpsM_41[,grep("^1",names(rbpsM_41), value = TRUE)]
T1F_WT = rbpsF_WT[,grep("^1",names(rbpsF_WT), value = TRUE)]
T1M_WT = rbpsM_WT[,grep("^1",names(rbpsM_WT), value = TRUE)]
T1F_36 = rbpsF_36[,grep("^1",names(rbpsF_36), value = TRUE)]
T1M_36 = rbpsM_36[,grep("^1",names(rbpsM_36), value = TRUE)]
T1F_TG = rbpsF_TG[,grep("^1",names(rbpsF_TG), value = TRUE)]
T1M_TG = rbpsM_TG[,grep("^1",names(rbpsM_TG), value = TRUE)]

T2F_41 = rbpsF_41[,grep("^2",names(rbpsF_41), value = TRUE)]
T2M_41 = rbpsM_41[,grep("^2",names(rbpsM_41), value = TRUE)]
T2F_WT = rbpsF_WT[,grep("^2",names(rbpsF_WT), value = TRUE)]
T2M_WT = rbpsM_WT[,grep("^2",names(rbpsM_WT), value = TRUE)]
T2F_36 = rbpsF_36[,grep("^2",names(rbpsF_36), value = TRUE)]
T2M_36 = rbpsM_36[,grep("^2",names(rbpsM_36), value = TRUE)]
T2F_TG = rbpsF_TG[,grep("^2",names(rbpsF_TG), value = TRUE)]
T2M_TG = rbpsM_TG[,grep("^2",names(rbpsM_TG), value = TRUE)]

T3F_41 = rbpsF_41[,grep("^3",names(rbpsF_41), value = TRUE)]
T3M_41 = rbpsM_41[,grep("^3",names(rbpsM_41), value = TRUE)]
T3F_WT = rbpsF_WT[,grep("^3",names(rbpsF_WT), value = TRUE)]
T3M_WT = rbpsM_WT[,grep("^3",names(rbpsM_WT), value = TRUE)]
T3F_36 = rbpsF_36[,grep("^3",names(rbpsF_36), value = TRUE)]
T3M_36 = rbpsM_36[,grep("^3",names(rbpsM_36), value = TRUE)]
T3F_TG = rbpsF_TG[,grep("^3",names(rbpsF_TG), value = TRUE)]
T3M_TG = rbpsM_TG[,grep("^3",names(rbpsM_TG), value = TRUE)]

#Binding final objects for comparison between different conditions (obj names = Timepoint-Sex-Line-conditions)
T1F_WT36 = bind_cols(T1F_WT, T1F_36)
T1M_WT36 = bind_cols(T1M_WT, T1M_36)
T1F_WT41 = bind_cols(T1F_WT, T1F_41)
T1M_WT41 = bind_cols(T1M_WT, T1M_41)
T1F_3641 = bind_cols(T1F_36, T1F_41)
T1M_3641 = bind_cols(T1M_36, T1M_41)
T1F_WTG = bind_cols(T1F_WT, T1F_TG)
T1M_WTG = bind_cols(T1M_WT, T1M_TG)
  
T2F_WT36 = bind_cols(T2F_WT, T2F_36)
T2M_WT36 = bind_cols(T2M_WT, T2M_36)
T2F_WT41 = bind_cols(T2F_WT, T2F_41)
T2M_WT41 = bind_cols(T2M_WT, T2M_41)
T2F_3641 = bind_cols(T2F_36, T2F_41)
T2M_3641 = bind_cols(T2M_36, T2M_41)
T2F_WTG = bind_cols(T2F_WT, T2F_TG)
T2M_WTG = bind_cols(T2M_WT, T2M_TG)

  
T3F_WT41 = bind_cols(T3F_WT, T3F_41)
T3M_WT41 = bind_cols(T3M_WT, T3M_41)
T3F_WT36 = bind_cols(T3F_WT, T3F_36)
T3M_WT36 = bind_cols(T3M_WT, T3M_36)
T3F_3641 = bind_cols(T3F_36, T3F_41)
T3M_3641 = bind_cols(T3M_36, T3M_41)
T3F_WTG = bind_cols(T3F_WT, T3F_TG)
T3M_WTG = bind_cols(T3M_WT, T3M_TG)

```

## setting up dfs for DESeq design 
```{r}
#timepoint lists
tpF_WT = c("1","1","1","2","2","2","3","3","3")
tpM_WT = c("1","1","1","2","2","2","3","3","3")
tpF_TG = c("1","1","1","1","2","2","2","3","1","1","1","2","2","2","3","3")
tpM_TG = c("1","1","2","2","2","3","3","3","3","3","1","1","1","2","2","2","3","3","3","3")

tpF_36 = c("1","1","1","1","2","2","2","3")
tpM_36 = c("1","1","2","2","2","3","3","3","3","3")

tpF_41 = c("1","1","1","2","2","2","3","3")
tpM_41 = c("1","1","1","2","2","2","3","3","3","3")

#line list

l1F_3641 = c("36","36","36","36","41","41","41")
l1M_3641 = c("36","36","41","41","41")

l1F_WTG = c("WT","WT","WT","TG","TG","TG","TG","TG","TG","TG")
l1M_WTG = c("WT","WT","WT","TG","TG","TG","TG","TG")

l1F_WT36 = c("WT","WT","WT","36","36","36","36")
l1M_WT36 = c("WT","WT","WT","36","36")

l1F_WT41 = c("WT","WT","WT","41","41","41")
l1M_WT41 = c("WT","WT","WT","41","41","41")

l2F_WTG = c("WT","WT","WT","TG","TG","TG","TG","TG","TG")
l2M_WTG = c("WT","WT","WT","TG","TG","TG","TG","TG","TG")

l2F_WT36 = c("WT","WT","WT","36","36","36")
l2M_WT36 = c("WT","WT","WT","36","36","36")

l2F_WT41 = c("WT","WT","WT","41","41","41")
l2M_WT41 = c("WT","WT","WT","41","41","41")

l2F_3641 = c("36","36","36","41","41","41")
l2M_3641 = c("36","36","36","41","41","41")

l3F_WTG = c("WT","WT","WT","TG","TG","TG")
l3M_WTG = c("WT","WT","WT","TG","TG","TG","TG","TG","TG","TG","TG","TG")

l3F_WT36 = c("WT","WT","WT","36")
l3M_WT36 = c("WT","WT","WT","36","36","36","36","36")

l3F_WT41 = c("WT","WT","WT","41","41")
l3M_WT41 = c("WT","WT","WT","41","41","41","41")

l3F_3641 = c("36","41","41")
l3M_3641 = c("36","36","36","36","36","41","41","41", "41")

# Binding timepoints and line lists for each condition

c1F_WTG = as.data.frame(cbind(colnames(T1F_WTG),l1F_WTG))
c1M_WTG = as.data.frame(cbind(colnames(T1M_WTG),l1M_WTG))

c1F_WT36 = as.data.frame(cbind(colnames(T1F_WT36),l1F_WT36))
c1M_WT36 = as.data.frame(cbind(colnames(T1M_WT36),l1M_WT36))

c1F_WT41 = as.data.frame(cbind(colnames(T1F_WT41),l1F_WT41))
c1M_WT41 = as.data.frame(cbind(colnames(T1M_WT41),l1M_WT41))

c1F_3641 = as.data.frame(cbind(colnames(T1F_3641),l1F_3641))
c1M_3641 = as.data.frame(cbind(colnames(T1M_3641),l1M_3641))

c2F_WTG = as.data.frame(cbind(colnames(T2F_WTG),l2F_WTG))
c2M_WTG = as.data.frame(cbind(colnames(T2M_WTG),l2M_WTG))

c2F_WT36 = as.data.frame(cbind(colnames(T2F_WT36),l2F_WT36))
c2M_WT36 = as.data.frame(cbind(colnames(T2M_WT36),l2M_WT36))

c2F_WT41 = as.data.frame(cbind(colnames(T2F_WT41),l2F_WT41))
c2M_WT41 = as.data.frame(cbind(colnames(T2M_WT41),l2M_WT41))

c2F_3641 = as.data.frame(cbind(colnames(T2F_3641),l2F_3641))
c2M_3641 = as.data.frame(cbind(colnames(T2M_3641),l2M_3641))

c3F_WTG = as.data.frame(cbind(colnames(T3F_WTG),l3F_WTG))
c3M_WTG = as.data.frame(cbind(colnames(T3M_WTG),l3M_WTG))

c3F_WT36 = as.data.frame(cbind(colnames(T3F_WT36),l3F_WT36))
c3M_WT36 = as.data.frame(cbind(colnames(T3M_WT36),l3M_WT36))

c3F_WT41 = as.data.frame(cbind(colnames(T3F_WT41),l3F_WT41))
c3M_WT41 = as.data.frame(cbind(colnames(T3M_WT41),l3M_WT41))

c3F_3641 = as.data.frame(cbind(colnames(T3F_3641),l3F_3641))
c3M_3641 = as.data.frame(cbind(colnames(T3M_3641),l3M_3641))

cF_WT = as.data.frame(cbind(colnames(rbpsF_WT),tpF_WT))
cM_WT = as.data.frame(cbind(colnames(rbpsM_WT),tpM_WT))

cF_TG = as.data.frame(cbind(colnames(rbpsF_TG),tpF_TG))
cM_TG = as.data.frame(cbind(colnames(rbpsM_TG),tpM_TG))

cF_36 = as.data.frame(cbind(colnames(rbpsF_36),tpF_36))
cM_36 = as.data.frame(cbind(colnames(rbpsM_36),tpM_36))

cF_41 = as.data.frame(cbind(colnames(rbpsF_41), tpF_41))
cM_41 = as.data.frame(cbind(colnames(rbpsM_41), tpM_41))


```

## Conversion into DESeqDataSets

```{r message=FALSE, warning=FALSE, include=FALSE}

dds1F_WTG = DESeqDataSetFromMatrix(countData = T1F_WTG,colData = c1F_WTG,
                                  design = ~l1F_WTG)
dds1M_WTG = DESeqDataSetFromMatrix(countData = T1M_WTG,colData = c1M_WTG,
                                  design = ~l1M_WTG)

dds1F_WT36 = DESeqDataSetFromMatrix(countData = T1F_WT36, colData = c1F_WT36,
                                 design = ~l1F_WT36)
dds1M_WT36 = DESeqDataSetFromMatrix(countData = T1M_WT36, colData = c1M_WT36,
                                 design = ~l1M_WT36)

dds1F_WT41 = DESeqDataSetFromMatrix(countData = T1F_WT41,colData = c1F_WT41,
                                 design = ~l1F_WT41)
dds1M_WT41 = DESeqDataSetFromMatrix(countData = T1M_WT41,colData = c1M_WT41,
                                 design = ~l1M_WT41)

dds1F_3641 = DESeqDataSetFromMatrix(countData = T1F_3641,colData = c1F_3641,
                                    design = ~l1F_3641)
dds1M_3641 = DESeqDataSetFromMatrix(countData = T1M_3641,colData = c1M_3641,
                                    design = ~l1M_3641)

dds2F_WTG = DESeqDataSetFromMatrix(countData = T2F_WTG,colData = c2F_WTG,
                                  design = ~l2F_WTG)
dds2M_WTG = DESeqDataSetFromMatrix(countData = T2M_WTG,colData = c2M_WTG,
                                  design = ~l2M_WTG)

dds2F_WT36 = DESeqDataSetFromMatrix(countData = T2F_WT36,colData = c2F_WT36,
                                 design = ~l2F_WT36)
dds2M_WT36 = DESeqDataSetFromMatrix(countData = T2M_WT36,colData = c2M_WT36,
                                 design = ~l2M_WT36)

dds2F_WT41 = DESeqDataSetFromMatrix(countData = T2F_WT41,colData = c2F_WT41,
                                 design = ~l2F_WT41)
dds2M_WT41 = DESeqDataSetFromMatrix(countData = T2M_WT41,colData = c2M_WT41,
                                 design = ~l2M_WT41)

dds2F_3641 = DESeqDataSetFromMatrix(countData = T2F_3641,colData = c2F_3641,
                                    design = ~l2F_3641)
dds2M_3641 = DESeqDataSetFromMatrix(countData = T2M_3641,colData = c2M_3641,
                                    design = ~l2M_3641)

dds3F_WTG = DESeqDataSetFromMatrix(countData = T3F_WTG,colData = c3F_WTG,
                                  design = ~l3F_WTG)
dds3M_WTG = DESeqDataSetFromMatrix(countData = T3M_WTG,colData = c3M_WTG,
                                  design = ~l3M_WTG)

dds3F_WT36 = DESeqDataSetFromMatrix(countData = T3F_WT36,colData = c3F_WT36,
                                 design = ~l3F_WT36)
dds3M_WT36 = DESeqDataSetFromMatrix(countData = T3M_WT36,colData = c3M_WT36,
                                 design = ~l3M_WT36)

dds3F_WT41 = DESeqDataSetFromMatrix(countData = T3F_WT41,colData = c3F_WT41,
                                 design = ~l3F_WT41)
dds3M_WT41 = DESeqDataSetFromMatrix(countData = T3M_WT41, colData = c3M_WT41,
                                 design = ~l3M_WT41)

dds3F_3641 = DESeqDataSetFromMatrix(countData = T3F_3641,colData = c3F_3641,
                                    design = ~l3F_3641)
dds3M_3641 = DESeqDataSetFromMatrix(countData = T3M_3641,colData = c3M_3641,
                                    design = ~l3M_3641)

ddsF_WT = DESeqDataSetFromMatrix(countData = rbpsF_WT,colData = cF_WT,
                                design = ~tpF_WT)
ddsM_WT = DESeqDataSetFromMatrix(countData = rbpsM_WT,colData = cM_WT,
                                design = ~tpM_WT)

ddsF_TG = DESeqDataSetFromMatrix(countData = rbpsF_TG,colData = cF_TG,
                                design = ~tpF_TG)
ddsM_TG = DESeqDataSetFromMatrix(countData = rbpsM_TG,colData = cM_TG,
                                design = ~tpM_TG)

ddsF_36 = DESeqDataSetFromMatrix(countData = rbpsF_36,colData = cF_36,
                                design = ~tpF_36)
ddsM_36 = DESeqDataSetFromMatrix(countData = rbpsM_36,colData = cM_36,
                                design = ~tpM_36)

ddsF_41 = DESeqDataSetFromMatrix(countData = rbpsF_41,colData = cF_41,
                                design = ~tpF_41)
ddsM_41 = DESeqDataSetFromMatrix(countData = rbpsM_41,colData = cM_41,
                                design = ~tpM_41)
```

## Running DESeq analysis

```{r message=FALSE}

dds1F_WTG = DESeq(dds1F_WTG)
dds1M_WTG = DESeq(dds1M_WTG)

dds1F_WT36 = DESeq(dds1F_WT36)
dds1M_WT36 = DESeq(dds1M_WT36)

dds1F_WT41 = DESeq(dds1F_WT41)
dds1M_WT41 = DESeq(dds1M_WT41)

dds1F_3641 = DESeq(dds1F_3641)
dds1M_3641 = DESeq(dds1M_3641)

dds2F_WTG = DESeq(dds2F_WTG)
dds2M_WTG = DESeq(dds2M_WTG)

dds2F_WT36 = DESeq(dds2F_WT36)
dds2M_WT36 = DESeq(dds2M_WT36)

dds2F_WT41 = DESeq(dds2F_WT41)
dds2M_WT41 = DESeq(dds2M_WT41)

dds2F_3641 = DESeq(dds2F_3641)
dds2M_3641 = DESeq(dds2M_3641)

dds3F_WTG = DESeq(dds3F_WTG)
dds3M_WTG = DESeq(dds3M_WTG)

dds3F_WT36 = DESeq(dds3F_WT36)
dds3M_WT36 = DESeq(dds3M_WT36)

dds3F_WT41 = DESeq(dds3F_WT41)
dds3M_WT41 = DESeq(dds3M_WT41)

dds3F_3641 = DESeq(dds3F_3641)
dds3M_3641 = DESeq(dds3M_3641)

ddsF_WT = DESeq(ddsF_WT)
ddsM_WT = DESeq(ddsM_WT)

ddsF_TG = DESeq(ddsF_TG)
ddsM_TG = DESeq(ddsM_TG)

ddsF_36 = DESeq(ddsF_36)
ddsM_36 = DESeq(ddsM_36)

ddsF_41 = DESeq(ddsF_41)
ddsM_41 = DESeq(ddsM_41)
```

# Preparing readable output

## Converting to result objects

```{r}

res1F_WTG = results(dds1F_WTG, contrast = c("l1F_WTG", "TG", "WT"))
res1M_WTG = results(dds1M_WTG, contrast = c("l1M_WTG", "TG", "WT"))

res1F_WT36 = results(dds1F_WT36, contrast = c("l1F_WT36", "36", "WT"))
res1M_WT36 = results(dds1M_WT36, contrast = c("l1M_WT36", "36", "WT"))

res1F_WT41 = results(dds1F_WT41, contrast = c("l1F_WT41", "41", "WT"))
res1M_WT41 = results(dds1M_WT41, contrast = c("l1M_WT41", "41", "WT"))

res1F_3641 = results(dds1F_3641, contrast = c("l1F_3641", "41", "36"))
res1M_3641 = results(dds1M_3641, contrast = c("l1M_3641", "41", "36"))

res2F_WTG = results(dds2F_WTG, contrast = c("l2F_WTG", "TG", "WT"))
res2M_WTG = results(dds2M_WTG, contrast = c("l2M_WTG", "TG", "WT"))

res2F_WT36 = results(dds2F_WT36, contrast = c("l2F_WT36", "36", "WT"))
res2M_WT36 = results(dds2M_WT36, contrast = c("l2M_WT36", "36", "WT"))

res2F_WT41 = results(dds2F_WT41, contrast = c("l2F_WT41", "41", "WT"))
res2M_WT41 = results(dds2M_WT41, contrast = c("l2M_WT41", "41", "WT"))

res2F_3641 = results(dds2F_3641, contrast = c("l2F_3641", "41", "36"))
res2M_3641 = results(dds2M_3641, contrast = c("l2M_3641", "41", "36"))

res3F_WTG = results(dds3F_WTG, contrast = c("l3F_WTG", "TG", "WT"))
res3M_WTG = results(dds3M_WTG, contrast = c("l3M_WTG", "TG", "WT"))

res3F_WT36 = results(dds3F_WT36, contrast = c("l3F_WT36", "36", "WT"))
res3M_WT36 = results(dds3M_WT36, contrast = c("l3M_WT36", "36", "WT"))

res3F_WT41 = results(dds3F_WT41, contrast = c("l3F_WT41", "41", "WT"))
res3M_WT41 = results(dds3M_WT41, contrast = c("l3M_WT41", "41", "WT"))

res3F_3641 = results(dds3F_3641, contrast = c("l3F_3641","41", "36"))
res3M_3641 = results(dds3M_3641, contrast = c("l3M_3641","41", "36"))
                    
resF_WT31 = results(ddsF_WT, contrast = c("tpF_WT","3", "1"))
resF_WT21 = results(ddsF_WT, contrast = c("tpF_WT","2", "1"))
resF_WT32 = results(ddsF_WT, contrast = c("tpF_WT","3", "2"))

resM_WT31 = results(ddsM_WT, contrast = c("tpM_WT","3","1"))
resM_WT21 = results(ddsM_WT, contrast = c("tpM_WT","2","1"))
resM_WT32 = results(ddsM_WT, contrast = c("tpM_WT","3","2"))

resF_TG31 = results(ddsF_TG, contrast = c("tpF_TG","3","1"))
resF_TG21 = results(ddsF_TG, contrast = c("tpF_TG","2","1"))
resF_TG32 = results(ddsF_TG, contrast = c("tpF_TG","3","2"))

resM_TG31 = results(ddsM_TG, contrast = c("tpM_TG","3","1"))
resM_TG21 = results(ddsM_TG, contrast = c("tpM_TG","2","1"))
resM_TG32 = results(ddsM_TG, contrast = c("tpM_TG","3","2"))

resF_36_31 = results(ddsF_36, contrast = c("tpF_36","3","1"))
resF_36_21 = results(ddsF_36, contrast = c("tpF_36","2","1"))
resF_36_32 = results(ddsF_36, contrast = c("tpF_36","3","2"))

resM_36_31 = results(ddsM_36, contrast = c("tpM_36","3","1"))
resM_36_21 = results(ddsM_36, contrast = c("tpM_36","2","1"))
resM_36_32 = results(ddsM_36, contrast = c("tpM_36","3","2"))

resF_41_31 = results(ddsF_41, contrast = c("tpF_41","3","1"))
resF_41_21 = results(ddsF_41, contrast = c("tpF_41","2","1"))
resF_41_32 = results(ddsF_41, contrast = c("tpF_41","3","2"))

resM_41_31 = results(ddsM_41, contrast = c("tpM_41","3","1"))
resM_41_21 = results(ddsM_41, contrast = c("tpM_41","2","1"))
resM_41_32 = results(ddsM_41, contrast = c("tpM_41","3","2"))

```

## Converting results to dfs, adding MGI,
entrez and Symbol as columns 

```{r message=FALSE}

# Making a list out of all results objects

allObj = ls()  

reObj = allObj[grep("^res", allObj)]

reList = mget(reObj)

# Adding back MGI, RefSeq ID and symbol columns to all results objects

reList = lapply(reList, function(res) {

  df = as.data.frame(res)

  df$MGI = rownames(df)
  
  df = merge(df, reads[, c("MGI", "Gene_ID", "Gene_Symbol")], by = "MGI", all.x = TRUE)
  
  rownames(df) <- NULL
  
  return(df)
})

colonies = list(reList$res1F_WT36, reList$res2F_WT36, reList$res3F_WT36, reList$res1M_WT36, reList$res2M_WT36, reList$res2M_WT36, reList$res1F_WT41, reList$res2F_WT41, reList$res3F_WT41, reList$res1M_WT41, reList$res2M_WT41, reList$res2M_WT41)
```

## Function for selecting genes according to LFC, padj value and key

```{r}

pdegenes = function(x, lfcThreshold = 2, padjThreshold = 0.05, max = 20, key = "Gene_Symbol") {
  
  # Filter rows based on log2FoldChange and padj
  filtered_x <- x[(x$log2FoldChange > lfcThreshold) & (x$padj < padjThreshold) & !is.na(x$padj), ]
  
  # Sort the filtered rows by padj value
  sorted_x <- filtered_x[order(filtered_x$padj), ]
  
  # Handle max condition. If a number, returned DF has that many lines
  if (!is.null(max)) {
    sorted_x <- head(sorted_x, max)
  }
  
  return(sorted_x[[key]])
}

ndegenes = function(x, lfcThreshold = -2, padjThreshold = 0.05, max = 20, key = "Gene_Symbol") {

  filtered_x <- x[(x$log2FoldChange < lfcThreshold) & (x$padj < padjThreshold) & !is.na(x$padj), ]
  
  sorted_x <- filtered_x[order(filtered_x$padj), ]
  
  if (!is.null(max)) {
    sorted_x <- head(sorted_x, max)
  }
  
  return(sorted_x[[key]])
}

degenes = function(x, lfcThreshold = 2, padjThreshold = 0.05, max = 20, key = "Gene_Symbol") {
  
  filtered_x <- x[(x$log2FoldChange > lfcThreshold | x$log2FoldChange < -lfcThreshold) & (x$padj < padjThreshold) & !is.na(x$padj),]
  
  sorted_x <- filtered_x[order(filtered_x$padj), ]
  
  if (!is.null(max)) {
    sorted_x <- head(sorted_x, max)
  }
  
  return(sorted_x[[key]])
}

```

## Applying to all results

```{r}

genespAll = lapply(reList, pdegenes, lfcThreshold = 0.5, padjThreshold = 0.5, max = NULL)

genesnAll = lapply(reList, ndegenes, lfcThreshold = -0.5, padjThreshold = 0.5, max = NULL)

genesAll = lapply(reList, degenes, lfcThreshold = 0.5, padjThreshold = 0.5, max = NULL)

```

## Outputting as csv for easy consulting

```{r message=FALSE, warning=FALSE}

outputTidy = function(x, title) {

  max_length = max(sapply(x, length))
  
  padded_lists = lapply(x, function(y) {
    length(y) = max_length  
    return(y)
  })
  
  # Convert the list of lists into a data frame
  df = as.data.frame(padded_lists)
  
  # Step 3: Write the data frame to a CSV file
  write.csv(df, title, row.names = FALSE)
  
  # Return the data frame (optional, for checking)
  return(df)
}

outputTidy(genesAll, "DEG_rbps.csv")
outputTidy(genespAll, "DEG_P_rbps.csv")
outputTidy(genesnAll, "DEG_N_rbps.csv")

#When opening the csv files in excel, navigate to data tab and click on "Text to columns" when selecting the entire first column, select "Delimited", add "comma" as a separator and format all columns as text. 


```


## Misc.

```{r}
medianpLFC = function(x){
  
  x = x[(x$log2FoldChange > 0) & (x$padj < 0.5),]
  #x = x[(x$log2FoldChange > 0.5) & (x$padj < 0.5),]
  
  y = median(x$log2FoldChange, na.rm = TRUE)
  
  return(y)
}

mediannLFC = function(x){

  x = x[(x$log2FoldChange < 0) & (x$padj < 0.5),]    
  #x = x[(x$log2FoldChange < -0.5) & (x$padj < 0.5),]
  
  y = median(x$log2FoldChange, na.rm = TRUE)
  
  return(y)
}
```

```{r}

T1F_WT36 = medianpLFC(reList$res1F_WT36)

T2F_WT36 = medianpLFC(reList$res2F_WT36)

T3F_WT36 = medianpLFC(reList$res3F_WT36)

T1M_WT36 = medianpLFC(reList$res1M_WT36)

T2M_WT36 = medianpLFC(reList$res2M_WT36)

T3M_WT36 = medianpLFC(reList$res3M_WT36)

medPlfc36 = c(T1F_WT36, T2F_WT36, T3F_WT36, T1M_WT36, T2M_WT36, T3M_WT36)

barplot(medPlfc36, names.arg = c("T1F", "T2F", "T3F", "T1M", "T2M", "T3M"), main = "Median Expression Log Fold Change for L36 PDEs", ylab = "Median Expression Log Fold Change")
```

```{r}
T1F_WT41 = medianpLFC(reList$res1F_WT41)

T2F_WT41 = medianpLFC(reList$res2F_WT41)

T3F_WT41 = medianpLFC(reList$res3F_WT41)

T1M_WT41 = medianpLFC(reList$res1M_WT41)

T2M_WT41 = medianpLFC(reList$res2M_WT41)

T3M_WT41 = medianpLFC(reList$res3M_WT41)

medPlfc41 = c(T1F_WT41, T2F_WT41, T3F_WT41, T1M_WT41, T2M_WT41, T3M_WT41)

barplot(medPlfc41, names.arg = c("T1F", "T2F", "T3F", "T1M", "T2M", "T3M"), main = "Median Expression Log Fold Change for L41 PDEs", ylab = "Average Expression Log Fold Change")
```

```{r}
T1F_WT36 = mediannLFC(reList$res1F_WT36)

T2F_WT36 = mediannLFC(reList$res2F_WT36)

T3F_WT36 = mediannLFC(reList$res3F_WT36)

T1M_WT36 = mediannLFC(reList$res1M_WT36)

T2M_WT36 = mediannLFC(reList$res2M_WT36)

T3M_WT36 = mediannLFC(reList$res3M_WT36)

medNlfc36 = c(T1F_WT36, T2F_WT36, T3F_WT36, T1M_WT36, T2M_WT36, T3M_WT36)

barplot(medNlfc36, names.arg = c("T1F", "T2F", "T3F", "T1M", "T2M", "T3M"), main = "Median Expression Log Fold Change for L36 NDEs", ylab = "Median Expression Log Fold Change")
```


```{r}
T1F_WT41 = mediannLFC(reList$res1F_WT41)

T2F_WT41 = mediannLFC(reList$res2F_WT41)

T3F_WT41 = mediannLFC(reList$res3F_WT41)

T1M_WT41 = mediannLFC(reList$res1M_WT41)

T2M_WT41 = mediannLFC(reList$res2M_WT41)

T3M_WT41 = mediannLFC(reList$res3M_WT41)

medNlfc41 = c(T1F_WT41, T2F_WT41, T3F_WT41, T1M_WT41, T2M_WT41, T3M_WT41)

barplot(medNlfc41, names.arg = c("T1F", "T2F", "T3F", "T1M", "T2M", "T3M"), main = "Average Expression Log Fold Change for L41 NDEs", ylab = "Median Expression Log Fold Change")
```

```{r}
DEGplot = function(x){
  
  y = x[x$log2FoldChange > 0.5 & x$padj < 0.5]
  
  z = x[x$log2FoldChange < -0.5 & x$padj < 0.5]
  
  length(y)
  
  
} 
```

```{r}
plot_DEG_barplot <- function(condition_dfs, condition_names, logfc_threshold = 0.5, padj_threshold = 0.5, title = "title your stuff") {
  
  # Check if the input is a list of data frames
  if (!is.list(condition_dfs) || any(!sapply(condition_dfs, is.data.frame))) {
    stop("Error: condition_dfs must be a list of data frames.")
  }
  
  # Initialize vectors to store the number of positive and negative DE genes for each condition
  positive_genes <- numeric(length(condition_dfs))
  negative_genes <- numeric(length(condition_dfs))
  
  # Loop over each data frame to count positive and negative DE genes
  for (i in seq_along(condition_dfs)) {
    df <- condition_dfs[[i]]
    
    # Ensure the data frame contains 'log2FoldChange' and 'pAdj' columns
    if (!all(c("log2FoldChange", "padj") %in% colnames(df))) {
      stop(paste("Error: Data frame for condition", condition_names[i], "must contain 'log2FoldChange' and 'pAdj' columns."))
    }
    
    # Count the number of positively and negatively differentially expressed genes
    positive_genes[i] <- sum(df$log2FoldChange > logfc_threshold & df$padj < padj_threshold, na.rm = TRUE)
    negative_genes[i] <- sum(df$log2FoldChange < -logfc_threshold & df$padj < padj_threshold, na.rm = TRUE)
  }
  
  # Combine the counts into a matrix for the barplot
  gene_counts <- rbind(negative_genes, positive_genes)
  
  # Calculate the range for the y-axis
  y_min <- min(c(0, negative_genes))  # Start from 0 to include positive values
  y_max <- max(c(positive_genes, abs(negative_genes))) * 1.2  # Extend the limit a bit to leave space for the labels
  
  # Plot the stacked barplot
  bar_positions <- barplot(gene_counts,
                           beside = TRUE,  # Plot bars side by side
                           names.arg = condition_names,  # Condition labels
                           col = c("red", "green"),  # Colors for positive and negative
                           main = title,
                           ylab = "Number of Genes",
                           ylim = c(y_min, y_max))  # Adjust the y-axis limits
  
  # Add a legend
  legend("topright", legend = c("Negatively DEGs", "Positively DEGs"), fill = c("red", "green"))
  
  # Add text on top of each bar showing the count
  text(x = bar_positions[1, ], y = negative_genes, labels = negative_genes, pos = 3, col = "black")  # Negative DEGs
  text(x = bar_positions[2, ], y = positive_genes, labels = positive_genes, pos = 3, col = "black")  # Positive DEGs
}
```

```{r}

L36 = list(reList$res1F_WT36, reList$res1M_WT36, reList$res2F_WT36, reList$res2M_WT36, reList$res3F_WT36, reList$res3M_WT36)

plot_DEG_barplot(condition_dfs = L36, condition_names = c("T1F", "T1M", "T2F", "T2M", "T3F", "T3M"), title = "Total Differentially Expressed Genes - L36")
```

```{r}
L41 = list(reList$res1F_WT41, reList$res1M_WT41, reList$res2F_WT41, reList$res2M_WT41, reList$res3F_WT41, reList$res3M_WT41)

plot_DEG_barplot(condition_dfs = L41, condition_names = c("T1F", "T1M", "T2F", "T2M", "T3F", "T3M"), title = "Total Differentially Expressed Genes - L41")
```


```{r, fig.height=6}

intersectTPs =  function(x,y){

  ysig = y[(y$log2FoldChange > 0.5 | y$log2FoldChange < -0.5) & (y$padj < 0.5),]
  
  xsig = x[(x$log2FoldChange > 0.5 | y$log2FoldChange < -0.5) & (x$padj < 0.5),]
  
  xsig = xsig[!(rownames(xsig) %in% rownames(ysig)),]
  
  return(xsig)
}

resF_36_21n = intersectTPs(x = reList$resF_36_21,y = reList$resF_WT21)
resM_36_21n = intersectTPs(x = reList$resM_36_21,y = reList$resM_WT21)

resF_36_32n = intersectTPs(x = reList$resF_36_32,y = reList$resF_WT32)
resM_36_32n = intersectTPs(x = reList$resM_36_32,y = reList$resM_WT32)

resF_36_31n = intersectTPs(x = reList$resF_36_31,y = reList$resF_WT31)
resM_36_31n = intersectTPs(x = reList$resM_36_31,y = reList$resM_WT31)


TP36F = list(reList$resF_36_21, resF_36_21n, reList$resF_36_32, resF_36_32n, reList$resF_36_31, resF_36_31n)

plot_DEG_barplot(condition_dfs = TP36F, condition_names = c("TP1-2", "TP1-2 -WT", "TP2-3", "TP2-3 -WT", "TP1-3", "TP1-3 -WT"), title = "Female L36 DEGs")

TP36M = list(reList$resM_36_21, resM_36_21n, reList$resM_36_32, resM_36_32n, reList$resM_36_31, resM_36_31n)

plot_DEG_barplot(condition_dfs = TP36M, condition_names = c("TP1-2", "TP1-2 -WT", "TP2-3", "TP2-3 -WT", "TP1-3", "TP1-3 -WT"), title = "Male L36 DEGs")

```


```{r, fig.height=6}
resF_41_21n = intersectTPs(x = reList$resF_41_21,y = reList$resF_WT21)
resM_41_21n = intersectTPs(x = reList$resM_41_21,y = reList$resM_WT21)

resF_41_32n = intersectTPs(x = reList$resF_41_32,y = reList$resF_WT32)
resM_41_32n = intersectTPs(x = reList$resM_41_32,y = reList$resM_WT32)

resF_41_31n = intersectTPs(x = reList$resF_41_31,y = reList$resF_WT31)
resM_41_31n = intersectTPs(x = reList$resM_41_31,y = reList$resM_WT31)


TP41F = list(reList$resF_41_21, resF_41_21n, reList$resF_41_32, resF_41_32n, reList$resF_41_31, resF_41_31n)

plot_DEG_barplot(condition_dfs = TP41F, condition_names = c("TP1-2", "TP1-2 -WT", "TP2-3", "TP2-3 -WT", "TP1-3", "TP1-3 -WT"), title = "Female L41 DEGs")

TP41M = list(reList$resM_41_21, resM_41_21n, reList$resM_41_32, resM_41_32n, reList$resM_41_31, resM_41_31n)

plot_DEG_barplot(condition_dfs = TP41M, condition_names = c("TP1-2", "TP1-2 -WT", "TP2-3", "TP2-3 -WT", "TP1-3", "TP1-3 -WT"), title = "Male L41 DEGs")

```

```{r}
RBPall = read.csv(file = "DEG_rbps.csv",sep = ",")
```

```{r}
ratio = function(x){
  
  full = length(genesAll[[x]])
  
  rbp = length(RBPall[[x]])
  
  # Calculate the percentage of rbp in relation to full
  rbp_percentage = (rbp / full) * 100
  
  # Create a pie chart with the ratio of full to rbp
  pie_values = c(full - rbp, rbp)
  labels = c(paste("Non-RBP", round(100 - rbp_percentage,2), "%"), 
             paste("RBP", round(rbp_percentage,2), "%"))
  
  pie(pie_values, labels = labels, main = paste("Ratio of", x, "in genesAll and RBPall"))
}


ratio(x = "res1F_WT36")
ratio(x = "res1M_WT36")

```
```{r}
ratio("res2F_WT36")
ratio("res2M_WT36")

```

```{r}
ratio("res3F_WT36")
ratio("res3M_WT36")

```

```{r}
ratio("res1F_WT41")
ratio("res1M_WT41")

```

```{r}
ratio("res2F_WT41")
ratio("res2M_WT41")
```

```{r}
ratio("res3F_WT41")
ratio("res3M_WT41")
```

```{r}
ratio("resF_36_21")
ratio("resM_36_21")
ratio("resF_41_21")
ratio("resM_41_21")
ratio("resF_WT21")
ratio("resM_WT21")
```

```{r}
ratio("resF_36_32")
ratio("resM_36_32")
ratio("resF_41_32")
ratio("resM_41_32")
ratio("resF_WT32")
ratio("resM_WT32")
```

```{r}
ratio("resF_36_31")
ratio("resM_36_31")
ratio("resF_41_31")
ratio("resM_41_31")
ratio("resF_WT31")
ratio("resM_WT31")
```

```{r}
PC_markers = read_excel("pc_markers.xlsx",col_names = FALSE)

res1F_WT36[reList$res1F_WT36$MGI %in% PC_markers$...1,]

pc_mark = function(x){
  
  # Ensure 'MGI' column exists in the DESeq2 results object (x)
  if (!"MGI" %in% colnames(x)) {
    stop("Error: 'MGI' column not found in the DESeq2 results object")
  }
  
  # Subset 'x' where 'MGI' is present in the PC_markers column
  x_subset = x[x$MGI %in% PC_markers[[1]], ]
  
  x_subset = (x_subset[(x_subset$log2FoldChange > 0.5 | x_subset$log2FoldChange < -0.5) & x_subset$padj < 0.05,])
  
  x_subset = na.omit(x_subset)
  
  return(x_subset)
}
```

```{r}
pc_mark(reList$res1F_WT36)
pc_mark(reList$res1M_WT36)

pc_mark(reList$res2F_WT36)
pc_mark(reList$res2M_WT36)

pc_mark(reList$res3F_WT36)
pc_mark(reList$res3M_WT36)

pc_mark(reList$res1F_WT41)
pc_mark(reList$res1M_WT41)

pc_mark(reList$res2F_WT41)
pc_mark(reList$res2M_WT41)

pc_mark(reList$res3F_WT41)
pc_mark(reList$res3M_WT41)

```

