---
title: "Visualizations & Plots"
author: "Lorenzo.MCM"
date: "2024-07-19"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Initial Setup

## Installing and loading packages.

```{r eval=FALSE}

install.packages(c('dplyr','readxl','gplots','DESeq2'), repos = "http://cran.us.r-project.org" )

BiocManager::install('EnhancedVolcano')
```

```{r message=FALSE}
library(readxl)
library(DESeq2)
library(gplots)
library(dplyr)
library(EnhancedVolcano)
```

## Importing data and tidying up resulting data frame so rows are ID'd by Gene Symbol.NAs omitted for comparison with Gene list

```{r message=FALSE, warning=FALSE}

# Setting Working Directory to the folder which contains this .Rmd file

setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

# Importing main reads file

data = as.data.frame(read_excel("Reads_genes.xlsx"))

# Keeping only the columns with MGI, RefSeq ID and Gene symbols + raw counts, omitting NAs 

reads = data %>% select(MGI, Gene_ID, Gene_Symbol, contains("Count"))

reads = as.data.frame(reads)

reads = na.omit(reads)

```

## Importing panther-db list of RBPs and tidying it up (browsed for RNA Metabolism Proteins and filtered for Mus musculus, exported to and edited in excel so the only row was MGI IDs, removed genes that did not have MGI IDs)

```{r message=FALSE, warning=FALSE}

# Importing list of RBPs, omitting NAs

#mainPanther = as.data.frame(read_excel('rna_met.xlsx', col_names = FALSE))

mainPanther = as.data.frame(read_excel('Panther-RBPs.xlsx', col_names = FALSE))

mainPanther = mainPanther[!is.na(mainPanther)]

# Filtering the reads object so it only contains genes whose MGI is also present in the list of RBPs, resulting in a final "rbps" object 

rbps = reads[reads$MGI %in% mainPanther,]


```

## Naming conditions and final set up of rbps object

```{r message=FALSE, warning=FALSE}

# Naming rows in the rbps object after their MGI ID and removing all columns except those with the counts for each condition

rownames(rbps) = rbps$Gene_Symbol

rbps = rbps[,-c(1,2,3)]

# Naming raw counts columns after their conditions, based on an external file

colnames(rbps) = (read_excel("conditions.xlsx", col_names = FALSE))[1,]
              
```

# Setting up for DESeq

## Intermediate objects for selecting conditions for comparisons through DESeq

```{r message=FALSE, warning=FALSE}

#Splitting conditions for line and sex
rbpsF_WT = rbps %>% select(contains("FWT"))
rbpsM_WT = rbps %>% select(contains("MWT"))
rbpsF_TG = rbps %>% select(contains("F36"), contains("F41"))
rbpsM_TG = rbps %>% select(contains("M36"), contains("M41"))
rbpsF_36 = rbps %>% select(contains("F36"))
rbpsM_36 = rbps %>% select(contains("M36"))
rbpsF_41 = rbps %>% select(contains("F41"))
rbpsM_41 = rbps %>% select(contains("M41"))

#Final split into timepoints
T1F_41 = rbpsF_41[,grep("^1",names(rbpsF_41), value = TRUE)]
T1M_41 = rbpsM_41[,grep("^1",names(rbpsM_41), value = TRUE)]
T1F_WT = rbpsF_WT[,grep("^1",names(rbpsF_WT), value = TRUE)]
T1M_WT = rbpsM_WT[,grep("^1",names(rbpsM_WT), value = TRUE)]
T1F_36 = rbpsF_36[,grep("^1",names(rbpsF_36), value = TRUE)]
T1M_36 = rbpsM_36[,grep("^1",names(rbpsM_36), value = TRUE)]
T1F_TG = rbpsF_TG[,grep("^1",names(rbpsF_TG), value = TRUE)]
T1M_TG = rbpsM_TG[,grep("^1",names(rbpsM_TG), value = TRUE)]

T2F_41 = rbpsF_41[,grep("^2",names(rbpsF_41), value = TRUE)]
T2M_41 = rbpsM_41[,grep("^2",names(rbpsM_41), value = TRUE)]
T2F_WT = rbpsF_WT[,grep("^2",names(rbpsF_WT), value = TRUE)]
T2M_WT = rbpsM_WT[,grep("^2",names(rbpsM_WT), value = TRUE)]
T2F_36 = rbpsF_36[,grep("^2",names(rbpsF_36), value = TRUE)]
T2M_36 = rbpsM_36[,grep("^2",names(rbpsM_36), value = TRUE)]
T2F_TG = rbpsF_TG[,grep("^2",names(rbpsF_TG), value = TRUE)]
T2M_TG = rbpsM_TG[,grep("^2",names(rbpsM_TG), value = TRUE)]

T3F_41 = rbpsF_41[,grep("^3",names(rbpsF_41), value = TRUE)]
T3M_41 = rbpsM_41[,grep("^3",names(rbpsM_41), value = TRUE)]
T3F_WT = rbpsF_WT[,grep("^3",names(rbpsF_WT), value = TRUE)]
T3M_WT = rbpsM_WT[,grep("^3",names(rbpsM_WT), value = TRUE)]
T3F_36 = rbpsF_36[,grep("^3",names(rbpsF_36), value = TRUE)]
T3M_36 = rbpsM_36[,grep("^3",names(rbpsM_36), value = TRUE)]
T3F_TG = rbpsF_TG[,grep("^3",names(rbpsF_TG), value = TRUE)]
T3M_TG = rbpsM_TG[,grep("^3",names(rbpsM_TG), value = TRUE)]

#Binding final objects for comparison between different conditions (obj names = Timepoint-Sex-Line-conditions)
T1F_WT36 = bind_cols(T1F_WT, T1F_36)
T1M_WT36 = bind_cols(T1M_WT, T1M_36)
T1F_WT41 = bind_cols(T1F_WT, T1F_41)
T1M_WT41 = bind_cols(T1M_WT, T1M_41)
T1F_3641 = bind_cols(T1F_36, T1F_41)
T1M_3641 = bind_cols(T1M_36, T1M_41)
T1F_WTG = bind_cols(T1F_WT, T1F_TG)
T1M_WTG = bind_cols(T1M_WT, T1M_TG)
  
T2F_WT36 = bind_cols(T2F_WT, T2F_36)
T2M_WT36 = bind_cols(T2M_WT, T2M_36)
T2F_WT41 = bind_cols(T2F_WT, T2F_41)
T2M_WT41 = bind_cols(T2M_WT, T2M_41)
T2F_3641 = bind_cols(T2F_36, T2F_41)
T2M_3641 = bind_cols(T2M_36, T2M_41)
T2F_WTG = bind_cols(T2F_WT, T2F_TG)
T2M_WTG = bind_cols(T2M_WT, T2M_TG)

  
T3F_WT41 = bind_cols(T3F_WT, T3F_41)
T3M_WT41 = bind_cols(T3M_WT, T3M_41)
T3F_WT36 = bind_cols(T3F_WT, T3F_36)
T3M_WT36 = bind_cols(T3M_WT, T3M_36)
T3F_3641 = bind_cols(T3F_36, T3F_41)
T3M_3641 = bind_cols(T3M_36, T3M_41)
T3F_WTG = bind_cols(T3F_WT, T3F_TG)
T3M_WTG = bind_cols(T3M_WT, T3M_TG)

```

## setting up dfs for DESeq design

```{r}
#timepoint lists
tpF_WT = c("1","1","1","2","2","2","3","3","3")
tpM_WT = c("1","1","1","2","2","2","3","3","3")
tpF_TG = c("1","1","1","1","2","2","2","3","1","1","1","2","2","2","3","3")
tpM_TG = c("1","1","2","2","2","3","3","3","3","3","1","1","1","2","2","2","3","3","3","3")

tpF_36 = c("1","1","1","1","2","2","2","3")
tpM_36 = c("1","1","2","2","2","3","3","3","3","3")

tpF_41 = c("1","1","1","2","2","2","3","3")
tpM_41 = c("1","1","1","2","2","2","3","3","3","3")

#line list

l1F_3641 = c("36","36","36","36","41","41","41")
l1M_3641 = c("36","36","41","41","41")

l1F_WTG = c("WT","WT","WT","TG","TG","TG","TG","TG","TG","TG")
l1M_WTG = c("WT","WT","WT","TG","TG","TG","TG","TG")

l1F_WT36 = c("WT","WT","WT","36","36","36","36")
l1M_WT36 = c("WT","WT","WT","36","36")

l1F_WT41 = c("WT","WT","WT","41","41","41")
l1M_WT41 = c("WT","WT","WT","41","41","41")

l2F_WTG = c("WT","WT","WT","TG","TG","TG","TG","TG","TG")
l2M_WTG = c("WT","WT","WT","TG","TG","TG","TG","TG","TG")

l2F_WT36 = c("WT","WT","WT","36","36","36")
l2M_WT36 = c("WT","WT","WT","36","36","36")

l2F_WT41 = c("WT","WT","WT","41","41","41")
l2M_WT41 = c("WT","WT","WT","41","41","41")

l2F_3641 = c("36","36","36","41","41","41")
l2M_3641 = c("36","36","36","41","41","41")

l3F_WTG = c("WT","WT","WT","TG","TG","TG")
l3M_WTG = c("WT","WT","WT","TG","TG","TG","TG","TG","TG","TG","TG","TG")

l3F_WT36 = c("WT","WT","WT","36")
l3M_WT36 = c("WT","WT","WT","36","36","36","36","36")

l3F_WT41 = c("WT","WT","WT","41","41")
l3M_WT41 = c("WT","WT","WT","41","41","41","41")

l3F_3641 = c("36","41","41")
l3M_3641 = c("36","36","36","36","36","41","41","41", "41")

# Binding timepoints and line lists for each condition

c1F_WTG = as.data.frame(cbind(colnames(T1F_WTG),l1F_WTG))
c1M_WTG = as.data.frame(cbind(colnames(T1M_WTG),l1M_WTG))

c1F_WT36 = as.data.frame(cbind(colnames(T1F_WT36),l1F_WT36))
c1M_WT36 = as.data.frame(cbind(colnames(T1M_WT36),l1M_WT36))

c1F_WT41 = as.data.frame(cbind(colnames(T1F_WT41),l1F_WT41))
c1M_WT41 = as.data.frame(cbind(colnames(T1M_WT41),l1M_WT41))

c1F_3641 = as.data.frame(cbind(colnames(T1F_3641),l1F_3641))
c1M_3641 = as.data.frame(cbind(colnames(T1M_3641),l1M_3641))

c2F_WTG = as.data.frame(cbind(colnames(T2F_WTG),l2F_WTG))
c2M_WTG = as.data.frame(cbind(colnames(T2M_WTG),l2M_WTG))

c2F_WT36 = as.data.frame(cbind(colnames(T2F_WT36),l2F_WT36))
c2M_WT36 = as.data.frame(cbind(colnames(T2M_WT36),l2M_WT36))

c2F_WT41 = as.data.frame(cbind(colnames(T2F_WT41),l2F_WT41))
c2M_WT41 = as.data.frame(cbind(colnames(T2M_WT41),l2M_WT41))

c2F_3641 = as.data.frame(cbind(colnames(T2F_3641),l2F_3641))
c2M_3641 = as.data.frame(cbind(colnames(T2M_3641),l2M_3641))

c3F_WTG = as.data.frame(cbind(colnames(T3F_WTG),l3F_WTG))
c3M_WTG = as.data.frame(cbind(colnames(T3M_WTG),l3M_WTG))

c3F_WT36 = as.data.frame(cbind(colnames(T3F_WT36),l3F_WT36))
c3M_WT36 = as.data.frame(cbind(colnames(T3M_WT36),l3M_WT36))

c3F_WT41 = as.data.frame(cbind(colnames(T3F_WT41),l3F_WT41))
c3M_WT41 = as.data.frame(cbind(colnames(T3M_WT41),l3M_WT41))

c3F_3641 = as.data.frame(cbind(colnames(T3F_3641),l3F_3641))
c3M_3641 = as.data.frame(cbind(colnames(T3M_3641),l3M_3641))

cF_WT = as.data.frame(cbind(colnames(rbpsF_WT),tpF_WT))
cM_WT = as.data.frame(cbind(colnames(rbpsM_WT),tpM_WT))

cF_TG = as.data.frame(cbind(colnames(rbpsF_TG),tpF_TG))
cM_TG = as.data.frame(cbind(colnames(rbpsM_TG),tpM_TG))

cF_36 = as.data.frame(cbind(colnames(rbpsF_36),tpF_36))
cM_36 = as.data.frame(cbind(colnames(rbpsM_36),tpM_36))

cF_41 = as.data.frame(cbind(colnames(rbpsF_41), tpF_41))
cM_41 = as.data.frame(cbind(colnames(rbpsM_41), tpM_41))


```

## Conversion into DESeqDataSets

```{r message=FALSE, warning=FALSE, include=FALSE}

dds1F_WTG = DESeqDataSetFromMatrix(countData = T1F_WTG,colData = c1F_WTG,
                                  design = ~l1F_WTG)
dds1M_WTG = DESeqDataSetFromMatrix(countData = T1M_WTG,colData = c1M_WTG,
                                  design = ~l1M_WTG)

dds1F_WT36 = DESeqDataSetFromMatrix(countData = T1F_WT36, colData = c1F_WT36,
                                 design = ~l1F_WT36)
dds1M_WT36 = DESeqDataSetFromMatrix(countData = T1M_WT36, colData = c1M_WT36,
                                 design = ~l1M_WT36)

dds1F_WT41 = DESeqDataSetFromMatrix(countData = T1F_WT41,colData = c1F_WT41,
                                 design = ~l1F_WT41)
dds1M_WT41 = DESeqDataSetFromMatrix(countData = T1M_WT41,colData = c1M_WT41,
                                 design = ~l1M_WT41)

dds1F_3641 = DESeqDataSetFromMatrix(countData = T1F_3641,colData = c1F_3641,
                                    design = ~l1F_3641)
dds1M_3641 = DESeqDataSetFromMatrix(countData = T1M_3641,colData = c1M_3641,
                                    design = ~l1M_3641)

dds2F_WTG = DESeqDataSetFromMatrix(countData = T2F_WTG,colData = c2F_WTG,
                                  design = ~l2F_WTG)
dds2M_WTG = DESeqDataSetFromMatrix(countData = T2M_WTG,colData = c2M_WTG,
                                  design = ~l2M_WTG)

dds2F_WT36 = DESeqDataSetFromMatrix(countData = T2F_WT36,colData = c2F_WT36,
                                 design = ~l2F_WT36)
dds2M_WT36 = DESeqDataSetFromMatrix(countData = T2M_WT36,colData = c2M_WT36,
                                 design = ~l2M_WT36)

dds2F_WT41 = DESeqDataSetFromMatrix(countData = T2F_WT41,colData = c2F_WT41,
                                 design = ~l2F_WT41)
dds2M_WT41 = DESeqDataSetFromMatrix(countData = T2M_WT41,colData = c2M_WT41,
                                 design = ~l2M_WT41)

dds2F_3641 = DESeqDataSetFromMatrix(countData = T2F_3641,colData = c2F_3641,
                                    design = ~l2F_3641)
dds2M_3641 = DESeqDataSetFromMatrix(countData = T2M_3641,colData = c2M_3641,
                                    design = ~l2M_3641)

dds3F_WTG = DESeqDataSetFromMatrix(countData = T3F_WTG,colData = c3F_WTG,
                                  design = ~l3F_WTG)
dds3M_WTG = DESeqDataSetFromMatrix(countData = T3M_WTG,colData = c3M_WTG,
                                  design = ~l3M_WTG)

dds3F_WT36 = DESeqDataSetFromMatrix(countData = T3F_WT36,colData = c3F_WT36,
                                 design = ~l3F_WT36)
dds3M_WT36 = DESeqDataSetFromMatrix(countData = T3M_WT36,colData = c3M_WT36,
                                 design = ~l3M_WT36)

dds3F_WT41 = DESeqDataSetFromMatrix(countData = T3F_WT41,colData = c3F_WT41,
                                 design = ~l3F_WT41)
dds3M_WT41 = DESeqDataSetFromMatrix(countData = T3M_WT41, colData = c3M_WT41,
                                 design = ~l3M_WT41)

dds3F_3641 = DESeqDataSetFromMatrix(countData = T3F_3641,colData = c3F_3641,
                                    design = ~l3F_3641)
dds3M_3641 = DESeqDataSetFromMatrix(countData = T3M_3641,colData = c3M_3641,
                                    design = ~l3M_3641)

ddsF_WT = DESeqDataSetFromMatrix(countData = rbpsF_WT,colData = cF_WT,
                                design = ~tpF_WT)
ddsM_WT = DESeqDataSetFromMatrix(countData = rbpsM_WT,colData = cM_WT,
                                design = ~tpM_WT)

ddsF_TG = DESeqDataSetFromMatrix(countData = rbpsF_TG,colData = cF_TG,
                                design = ~tpF_TG)
ddsM_TG = DESeqDataSetFromMatrix(countData = rbpsM_TG,colData = cM_TG,
                                design = ~tpM_TG)

ddsF_36 = DESeqDataSetFromMatrix(countData = rbpsF_36,colData = cF_36,
                                design = ~tpF_36)
ddsM_36 = DESeqDataSetFromMatrix(countData = rbpsM_36,colData = cM_36,
                                design = ~tpM_36)

ddsF_41 = DESeqDataSetFromMatrix(countData = rbpsF_41,colData = cF_41,
                                design = ~tpF_41)
ddsM_41 = DESeqDataSetFromMatrix(countData = rbpsM_41,colData = cM_41,
                                design = ~tpM_41)
```

## Running DESeq analysis

```{r message=FALSE}

dds1F_WTG = DESeq(dds1F_WTG)
dds1M_WTG = DESeq(dds1M_WTG)

dds1F_WT36 = DESeq(dds1F_WT36)
dds1M_WT36 = DESeq(dds1M_WT36)

dds1F_WT41 = DESeq(dds1F_WT41)
dds1M_WT41 = DESeq(dds1M_WT41)

dds1F_3641 = DESeq(dds1F_3641)
dds1M_3641 = DESeq(dds1M_3641)

dds2F_WTG = DESeq(dds2F_WTG)
dds2M_WTG = DESeq(dds2M_WTG)

dds2F_WT36 = DESeq(dds2F_WT36)
dds2M_WT36 = DESeq(dds2M_WT36)

dds2F_WT41 = DESeq(dds2F_WT41)
dds2M_WT41 = DESeq(dds2M_WT41)

dds2F_3641 = DESeq(dds2F_3641)
dds2M_3641 = DESeq(dds2M_3641)

dds3F_WTG = DESeq(dds3F_WTG)
dds3M_WTG = DESeq(dds3M_WTG)

dds3F_WT36 = DESeq(dds3F_WT36)
dds3M_WT36 = DESeq(dds3M_WT36)

dds3F_WT41 = DESeq(dds3F_WT41)
dds3M_WT41 = DESeq(dds3M_WT41)

dds3F_3641 = DESeq(dds3F_3641)
dds3M_3641 = DESeq(dds3M_3641)

ddsF_WT = DESeq(ddsF_WT)
ddsM_WT = DESeq(ddsM_WT)

ddsF_TG = DESeq(ddsF_TG)
ddsM_TG = DESeq(ddsM_TG)

ddsF_36 = DESeq(ddsF_36)
ddsM_36 = DESeq(ddsM_36)

ddsF_41 = DESeq(ddsF_41)
ddsM_41 = DESeq(ddsM_41)
```

# Preparing readable output

## Converting to result objects

```{r}

res1F_WTG = results(dds1F_WTG, contrast = c("l1F_WTG", "TG", "WT"))
res1M_WTG = results(dds1M_WTG, contrast = c("l1M_WTG", "TG", "WT"))

res1F_WT36 = results(dds1F_WT36, contrast = c("l1F_WT36", "36", "WT"))
res1M_WT36 = results(dds1M_WT36, contrast = c("l1M_WT36", "36", "WT"))

res1F_WT41 = results(dds1F_WT41, contrast = c("l1F_WT41", "41", "WT"))
res1M_WT41 = results(dds1M_WT41, contrast = c("l1M_WT41", "41", "WT"))

res1F_3641 = results(dds1F_3641, contrast = c("l1F_3641", "41", "36"))
res1M_3641 = results(dds1M_3641, contrast = c("l1M_3641", "41", "36"))

res2F_WTG = results(dds2F_WTG, contrast = c("l2F_WTG", "TG", "WT"))
res2M_WTG = results(dds2M_WTG, contrast = c("l2M_WTG", "TG", "WT"))

res2F_WT36 = results(dds2F_WT36, contrast = c("l2F_WT36", "36", "WT"))
res2M_WT36 = results(dds2M_WT36, contrast = c("l2M_WT36", "36", "WT"))

res2F_WT41 = results(dds2F_WT41, contrast = c("l2F_WT41", "41", "WT"))
res2M_WT41 = results(dds2M_WT41, contrast = c("l2M_WT41", "41", "WT"))

res2F_3641 = results(dds2F_3641, contrast = c("l2F_3641", "41", "36"))
res2M_3641 = results(dds2M_3641, contrast = c("l2M_3641", "41", "36"))

res3F_WTG = results(dds3F_WTG, contrast = c("l3F_WTG", "TG", "WT"))
res3M_WTG = results(dds3M_WTG, contrast = c("l3M_WTG", "TG", "WT"))

res3F_WT36 = results(dds3F_WT36, contrast = c("l3F_WT36", "36", "WT"))
res3M_WT36 = results(dds3M_WT36, contrast = c("l3M_WT36", "36", "WT"))

res3F_WT41 = results(dds3F_WT41, contrast = c("l3F_WT41", "41", "WT"))
res3M_WT41 = results(dds3M_WT41, contrast = c("l3M_WT41", "41", "WT"))

res3F_3641 = results(dds3F_3641, contrast = c("l3F_3641","41", "36"))
res3M_3641 = results(dds3M_3641, contrast = c("l3M_3641","41", "36"))
                    
resF_WT31 = results(ddsF_WT, contrast = c("tpF_WT","3", "1"))
resF_WT21 = results(ddsF_WT, contrast = c("tpF_WT","2", "1"))
resF_WT32 = results(ddsF_WT, contrast = c("tpF_WT","3", "2"))

resM_WT31 = results(ddsM_WT, contrast = c("tpM_WT","3","1"))
resM_WT21 = results(ddsM_WT, contrast = c("tpM_WT","2","1"))
resM_WT32 = results(ddsM_WT, contrast = c("tpM_WT","3","2"))

resF_TG31 = results(ddsF_TG, contrast = c("tpF_TG","3","1"))
resF_TG21 = results(ddsF_TG, contrast = c("tpF_TG","2","1"))
resF_TG32 = results(ddsF_TG, contrast = c("tpF_TG","3","2"))

resM_TG31 = results(ddsM_TG, contrast = c("tpM_TG","3","1"))
resM_TG21 = results(ddsM_TG, contrast = c("tpM_TG","2","1"))
resM_TG32 = results(ddsM_TG, contrast = c("tpM_TG","3","2"))

resF_36_31 = results(ddsF_36, contrast = c("tpF_36","3","1"))
resF_36_21 = results(ddsF_36, contrast = c("tpF_36","2","1"))
resF_36_32 = results(ddsF_36, contrast = c("tpF_36","3","2"))

resM_36_31 = results(ddsM_36, contrast = c("tpM_36","3","1"))
resM_36_21 = results(ddsM_36, contrast = c("tpM_36","2","1"))
resM_36_32 = results(ddsM_36, contrast = c("tpM_36","3","2"))

resF_41_31 = results(ddsF_41, contrast = c("tpF_41","3","1"))
resF_41_21 = results(ddsF_41, contrast = c("tpF_41","2","1"))
resF_41_32 = results(ddsF_41, contrast = c("tpF_41","3","2"))

resM_41_31 = results(ddsM_41, contrast = c("tpM_41","3","1"))
resM_41_21 = results(ddsM_41, contrast = c("tpM_41","2","1"))
resM_41_32 = results(ddsM_41, contrast = c("tpM_41","3","2"))

```

## Converting results to dfs, adding MGI,

entrez and Symbol as columns

```{r message=FALSE}

# Making a list out of all results objects

allObj = ls()  

reObj = allObj[grep("^res", allObj)]

reList = mget(reObj)

# Adding back MGI, RefSeq ID and symbol columns to all results objects

reList = lapply(reList, function(res) {

  df = as.data.frame(res)

  df$Gene_Symbol = rownames(df)
  
  df = merge(df, reads[, c("MGI", "Gene_ID", "Gene_Symbol")], by = "Gene_Symbol", all.x = TRUE)
  
  rownames(df) <- NULL
  
  return(df)
})

```

## Function for selecting genes according to LFC, padj value and key

```{r}

pdegenes = function(x, lfcThreshold = 2, padjThreshold = 0.05, max = 20, key = "Gene_Symbol") {
  
  # Filter rows based on log2FoldChange and padj
  filtered_x <- x[(x$log2FoldChange > lfcThreshold) & (x$padj < padjThreshold) & !is.na(x$padj), ]
  
  # Sort the filtered rows by LFC value
  sorted_x <- filtered_x[order(filtered_x$log2FoldChange), ]
  
  # Handle max condition. If a number, returned DF has that many lines
  if (!is.null(max)) {
    sorted_x <- head(sorted_x, max)
  }
  
  return(sorted_x[[key]])
}

ndegenes = function(x, lfcThreshold = -2, padjThreshold = 0.05, max = 20, key = "Gene_Symbol") {

  filtered_x <- x[(x$log2FoldChange < lfcThreshold) & (x$padj < padjThreshold) & !is.na(x$padj), ]
  
  sorted_x <- filtered_x[order(filtered_x$log2FoldChange), ]
  
  if (!is.null(max)) {
    sorted_x <- head(sorted_x, max)
  }
  
  return(sorted_x[[key]])
}

degenes = function(x, lfcThreshold = 2, padjThreshold = 0.05, max = 20, key = "Gene_Symbol") {
  
  filtered_x <- x[(x$log2FoldChange > lfcThreshold | x$log2FoldChange < -lfcThreshold) & (x$padj < padjThreshold) & !is.na(x$padj),]
  
  sorted_x <- filtered_x[order(filtered_x$log2FoldChange), ]
  
  if (!is.null(max)) {
    sorted_x <- head(sorted_x, max)
  }
  
  return(sorted_x[[key]])
}

```

## Applying to all results

```{r}

genespAll = lapply(reList, pdegenes, lfcThreshold = 0.5, padjThreshold = 0.05, max = NULL)

genesnAll = lapply(reList, ndegenes, lfcThreshold = -0.5, padjThreshold = 0.05, max = NULL)

genesAll = lapply(reList, degenes, lfcThreshold = 0.5, padjThreshold = 0.05, max = NULL)

```

# Visualizations

# rlog transform

```{r}
rl1F_WT36 = rlog(dds1F_WT36)
rl1M_WT36 = rlog(dds1M_WT36)

rl2F_WT36 = rlog(dds2F_WT36)
rl2M_WT36 = rlog(dds2M_WT36)

rl3F_WT36 = rlog(dds3F_WT36)
rl3M_WT36 = rlog(dds3M_WT36)

rl1F_WT41 = rlog(dds1F_WT41)
rl1M_WT41 = rlog(dds1M_WT41)

rl2F_WT41 = rlog(dds2F_WT41)
rl2M_WT41 = rlog(dds2M_WT41)

rl3F_WT41 = rlog(dds3F_WT41)
rl3M_WT41 = rlog(dds3M_WT41)

rl1F_3641 = rlog(dds1F_3641)
rl1M_3641 = rlog(dds1M_3641)

rl2F_3641 = rlog(dds2F_3641)
rl2M_3641 = rlog(dds2M_3641)

rl3F_3641 = rlog(dds3F_3641)
rl3M_3641 = rlog(dds3M_3641)

rlF_WT = rlog(ddsF_WT)
rlM_WT = rlog(ddsM_WT)

rlF_36 = rlog(ddsF_36)
rlM_36 = rlog(ddsM_36)

rlF_41 = rlog(ddsF_41)
rlM_41 = rlog(ddsM_41)


```

## PCA Plots

```{r}
plotPCA(rl1F_WT36, intgroup = c("l1F_WT36"))
plotPCA(rl1M_WT36, intgroup = c("l1M_WT36"))
```

```{r}
plotPCA(rl1F_WT41, intgroup = c("l1F_WT41"))
plotPCA(rl1M_WT41, intgroup = c("l1M_WT41"))
```

```{r}
plotPCA(rl1F_3641, intgroup = c("l1F_3641"))
plotPCA(rl1M_3641, intgroup = c("l1M_3641"))
```

```{r}
plotPCA(rl2F_WT36, intgroup = c("l2F_WT36"))
plotPCA(rl2M_WT36, intgroup = c("l2M_WT36"))
```

```{r}
plotPCA(rl2F_WT41, intgroup = c("l2F_WT41"))
plotPCA(rl2M_WT41, intgroup = c("l2M_WT41"))
```

```{r}
plotPCA(rl2F_3641, intgroup = c("l2F_3641"))
plotPCA(rl2M_3641, intgroup = c("l2M_3641"))
```

```{r}
plotPCA(rl3F_WT36, intgroup = c("l3F_WT36"))
plotPCA(rl3M_WT36, intgroup = c("l3M_WT36"))
```

```{r}
plotPCA(rl3F_WT41, intgroup = c("l3F_WT41"))
plotPCA(rl3M_WT41, intgroup = c("l3M_WT41"))
```

```{r}
plotPCA(rl3F_3641, intgroup = c("l3F_3641"))
plotPCA(rl3M_3641, intgroup = c("l3M_3641"))
```

```{r}
plotPCA(rlF_WT, intgroup = c("tpF_WT"))
plotPCA(rlM_WT, intgroup = c("tpM_WT"))
```

```{r}
plotPCA(rlF_36, intgroup = c("tpF_36"))
plotPCA(rlM_36, intgroup = c("tpM_36"))
```

```{r}
plotPCA(rlF_41, intgroup = c("tpF_41"))
plotPCA(rlM_41, intgroup = c("tpM_41"))
```

## Volcano Plots

### l36

```{r,fig.width= 7, fig.height=7}
library(EnhancedVolcano)
EnhancedVolcano(res1F_WT36,
                lab = rownames(res1F_WT36),
                x = 'log2FoldChange',
                y = 'pvalue',
                pCutoff = 0.05,
                pCutoffCol = 'padj',
                FCcutoff = 0.5,
                xlim = c(-6,6),
                title = NULL,
                subtitle = NULL,
                caption = NULL)
```

```{r,fig.width= 7, fig.height=7}
EnhancedVolcano(res1M_WT36,
                lab = rownames(res1M_WT36),
                x = 'log2FoldChange',
                y = 'pvalue',
                pCutoff = 0.05,
                pCutoffCol = 'padj',
                FCcutoff = 0.5,
                xlim = c(-3,3),
                title = NULL,
                subtitle = NULL,
                caption = NULL)
```

```{r,fig.width= 8, fig.height=8}
EnhancedVolcano(res2F_WT36,
                lab = rownames(res2F_WT36),
                x = 'log2FoldChange',
                y = 'pvalue',
                pCutoff = 0.05,
                pCutoffCol = 'padj',
                FCcutoff = 0.5,
                xlim = c(-3,3),
                title = NULL,
                subtitle = NULL,
                caption = NULL)
```

```{r,fig.width= 7, fig.height=7}
EnhancedVolcano(res2M_WT36,
                lab = rownames(res2M_WT36),
                x = 'log2FoldChange',
                y = 'pvalue',
                pCutoff = 0.05,
                pCutoffCol = 'padj',                
                FCcutoff = 0.5,
                xlim = c(-3,3),
                title = NULL,
                subtitle = NULL,
                caption = NULL,
                labSize = 5)

```

```{r,fig.width= 7, fig.height=7}
EnhancedVolcano(res3F_WT36,
                lab = rownames(res3F_WT36),
                x = 'log2FoldChange',
                y = 'pvalue',
                pCutoff = 0.05,
                pCutoffCol = 'padj',                
                FCcutoff = 0.5,
                xlim = c(-3,3),
                title = NULL,
                subtitle = NULL,
                caption = NULL)
```

```{r,fig.width= 7, fig.height=7}
EnhancedVolcano(res3M_WT36,
                lab = rownames(res3M_WT36),
                x = 'log2FoldChange',
                y = 'pvalue',
                pCutoff = 0.05,
                pCutoffCol = 'padj',                
                FCcutoff = 0.5,
                xlim = c(-3,3),
                title = NULL,
                subtitle = NULL,
                caption = NULL)
```

### l41

```{r,fig.width= 7, fig.height=7}
library(EnhancedVolcano)
EnhancedVolcano(res1F_WT41,
                lab = rownames(res1F_WT41),
                x = 'log2FoldChange',
                y = 'pvalue',
                pCutoff = 0.05,
                pCutoffCol = 'padj',
                FCcutoff = 0.5,
                xlim = c(-2,2),
                title = NULL,
                subtitle = NULL,
                caption = NULL)
```

```{r,fig.width= 7, fig.height=7}
EnhancedVolcano(res1M_WT41,
                lab = rownames(res1M_WT41),
                x = 'log2FoldChange',
                y = 'pvalue',
                pCutoff = 0.05,
                pCutoffCol = 'padj',
                FCcutoff = 0.5,
                xlim = c(-2,2),
                title = NULL,
                subtitle = NULL,
                caption = NULL)
```

```{r,fig.width= 7, fig.height=7}
EnhancedVolcano(res2F_WT41,
                lab = rownames(res2F_WT41),
                x = 'log2FoldChange',
                y = 'pvalue',
                pCutoff = 0.05,
                pCutoffCol = 'padj',
                FCcutoff = 0.5,
                xlim = c(-2,2),
                title = NULL,
                subtitle = NULL,
                caption = NULL)
```

```{r,fig.width= 7, fig.height=7}
EnhancedVolcano(res2M_WT41,
                lab = rownames(res2M_WT41),
                x = 'log2FoldChange',
                y = 'pvalue',
                pCutoff = 0.05,
                pCutoffCol = 'padj',                
                FCcutoff = 0.5,
                xlim = c(-3,2),
                title = NULL,
                subtitle = NULL,
                caption = NULL)
```

```{r,fig.width= 7, fig.height=7}
EnhancedVolcano(res3F_WT41,
                lab = rownames(res3F_WT41),
                x = 'log2FoldChange',
                y = 'pvalue',
                pCutoff = 0.05,
                pCutoffCol = 'padj',                
                FCcutoff = 0.5,
                xlim = c(-4,2),
                title = NULL,
                subtitle = NULL,
                caption = NULL)
```

```{r,fig.width= 7, fig.height=7}
EnhancedVolcano(res3M_WT41,
                lab = rownames(res3M_WT41),
                x = 'log2FoldChange',
                y = 'pvalue',
                pCutoff = 0.05,
                pCutoffCol = 'padj',                
                FCcutoff = 0.5,
                xlim = c(-3,3),
                title = NULL,
                subtitle = NULL,
                caption = NULL)
```

## Heatmaps

```{r}

detectGroups <- function (x){  # x are col names
  tem <- gsub("[0-9]*$","",x) # Remove all numbers from end
  #tem = gsub("_Rep|_rep|_REP","",tem)
  tem <- gsub("_$","",tem); # remove "_" from end
  tem <- gsub("_Rep$","",tem); # remove "_Rep" from end
  tem <- gsub("_rep$","",tem); # remove "_rep" from end
  tem <- gsub("_REP$","",tem)  # remove "_REP" from end
  return( tem )
}

dist2 <- function(x, ...)   # distance function = 1-PCC (Pearson's correlation coefficient)
  as.dist(1-cor(t(x), method="pearson"))

hclust2 <- function(x, method="average", ...)  # average linkage in hierarchical clustering
  hclust(x, method=method, ...)
```

```{r}
seqHeatmap = function(x,n = 50){
  
  x = assay(x)
  if(n>dim(x)[1]) n = dim(x)[1] # max as data
  
  x = x[order(apply(x,1,sd), decreasing = TRUE),] # sort genes by stardad deviation
  
  x = x[1:n,] # only keep the n genes
  
  # this will cutoff very large values, which could skew the color 
  x=as.matrix(x[1:n,])-apply(x[1:n,],1,mean)
  cutoff = median(unlist(x)) + 4*sd (unlist(x)) 
  x[x>cutoff] <- cutoff
  cutoff = median(unlist(x)) - 4*sd (unlist(x)) 
  x[x< cutoff] <- cutoff
  
  group = detectGroups(colnames(x))
  group.colors = rainbow(length(unique(group)))
  
  lmat = rbind(c(5,4), c(0,1), c(3,2))
  lwid = c(1.5,4)
  lhei = c(1,.2,4)
  
  par(mar = c(10, 10, 4, 2) + 0.1)  
  
  heatmap.2(x, distfun = dist2,hclustfun=hclust2,
	 col=greenred(75), density.info="none", trace="none", scale="none", keysize=.5
	,key=T, symkey=F
	,ColSideColors=group.colors[ as.factor(group)]
	,margins=c(8,12)
	,cexRow=1
	,srtCol=45
	,cexCol=1.  # size of font for sample names
	,lmat = lmat, lwid = lwid, lhei = lhei
	)
  
}

# Idk why but for this to output without an error (allowing knitting) you need to also add the figure dimensions at the start of the chunk
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rl1F_WT36)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rl1F_WT41)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rl1M_WT36)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rl1M_WT41)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rl2F_WT36)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rl2F_WT41)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rl2M_WT36)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rl2M_WT41)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rl3F_WT36)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rl3F_WT41)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rl3M_WT36)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rl3M_WT41)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rl1F_3641)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rl1M_3641)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rl2F_3641)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rl2M_3641)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rl3F_3641)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rl3M_3641)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rlF_36)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rlM_36)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rlF_41)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rlM_41)
```
```{r fig.height=8, fig.width=12}
seqHeatmap(rlF_WT)
```

```{r fig.height=8, fig.width=12}
seqHeatmap(rlM_WT)
```

## Median LFC for pAdj \< 0.05 genes

```{r}
medianpLFC = function(x){
  
  x = x[(x$log2FoldChange > 0) & (x$padj < 0.05),]
  #x = x[(x$log2FoldChange > 0.5) & (x$padj < 0.5),]
  
  y = median(x$log2FoldChange, na.rm = TRUE)
  
  return(y)
}

mediannLFC = function(x){

  x = x[(x$log2FoldChange < 0) & (x$padj < 0.05),]    
  #x = x[(x$log2FoldChange < -0.5) & (x$padj < 0.5),]
  
  y = median(x$log2FoldChange, na.rm = TRUE)
  
  return(y)
}
```

```{r}

T1F_WT36 = medianpLFC(reList$res1F_WT36)

T2F_WT36 = medianpLFC(reList$res2F_WT36)

T3F_WT36 = medianpLFC(reList$res3F_WT36)

T1M_WT36 = medianpLFC(reList$res1M_WT36)

T2M_WT36 = medianpLFC(reList$res2M_WT36)

T3M_WT36 = medianpLFC(reList$res3M_WT36)

medPlfc36 = c(T1F_WT36,T1M_WT36, T2F_WT36, T2M_WT36,T3F_WT36, T3M_WT36)

barplot(medPlfc36, names.arg = c("T1F", "T1M", "T2F", "T2M", "T3F", "T3M"), main = "Median Expression Log Fold Change for L36 PDEs", ylab = "Median Expression Log Fold Change")
```

```{r}
T1F_WT41 = medianpLFC(reList$res1F_WT41)

T2F_WT41 = medianpLFC(reList$res2F_WT41)

T3F_WT41 = medianpLFC(reList$res3F_WT41)

T1M_WT41 = medianpLFC(reList$res1M_WT41)

T2M_WT41 = medianpLFC(reList$res2M_WT41)

T3M_WT41 = medianpLFC(reList$res3M_WT41)

medPlfc41 = c(T1F_WT41,T1M_WT41, T2F_WT41, T2M_WT41,T3F_WT41, T3M_WT41)

barplot(medPlfc41, names.arg = c("T1F", "T1M", "T2F", "T2M", "T3F", "T3M"), main = "Median Expression Log Fold Change for L41 PDEs", ylab = "Average Expression Log Fold Change")
```

```{r}
T1F_WT36 = mediannLFC(reList$res1F_WT36)

T2F_WT36 = mediannLFC(reList$res2F_WT36)

T3F_WT36 = mediannLFC(reList$res3F_WT36)

T1M_WT36 = mediannLFC(reList$res1M_WT36)

T2M_WT36 = mediannLFC(reList$res2M_WT36)

T3M_WT36 = mediannLFC(reList$res3M_WT36)

medNlfc36 = c(T1F_WT36,T1M_WT36, T2F_WT36, T2M_WT36,T3F_WT36, T3M_WT36)

barplot(medNlfc36, names.arg = c("T1F", "T1M", "T2F", "T2M", "T3F", "T3M"), main = "Median Expression Log Fold Change for L36 NDEs", ylab = "Median Expression Log Fold Change")
```

```{r}
T1F_WT41 = mediannLFC(reList$res1F_WT41)

T2F_WT41 = mediannLFC(reList$res2F_WT41)

T3F_WT41 = mediannLFC(reList$res3F_WT41)

T1M_WT41 = mediannLFC(reList$res1M_WT41)

T2M_WT41 = mediannLFC(reList$res2M_WT41)

T3M_WT41 = mediannLFC(reList$res3M_WT41)

medNlfc41 = c(T1F_WT41,T1M_WT41, T2F_WT41, T2M_WT41,T3F_WT41, T3M_WT41)

barplot(medNlfc41, names.arg = c("T1F", "T1M", "T2F", "T2M", "T3F", "T3M"), main = "Average Expression Log Fold Change for L41 NDEs", ylab = "Median Expression Log Fold Change")
```

## Absolute numbers of RBP DEGs, split by condition

```{r}
plot_DEG_barplot <- function(condition_dfs, condition_names, logfc_threshold = 0.5, padj_threshold = 0.05, title = "title your stuff") {
  
  # Check if the input is a list of data frames
  if (!is.list(condition_dfs) || any(!sapply(condition_dfs, is.data.frame))) {
    stop("Error: condition_dfs must be a list of data frames.")
  }
  
  # Initialize vectors to store the number of positive and negative DE genes for each condition
  positive_genes <- numeric(length(condition_dfs))
  negative_genes <- numeric(length(condition_dfs))
  
  # Loop over each data frame to count positive and negative DE genes
  for (i in seq_along(condition_dfs)) {
    df <- condition_dfs[[i]]
    
    # Ensure the data frame contains 'log2FoldChange' and 'pAdj' columns
    if (!all(c("log2FoldChange", "padj") %in% colnames(df))) {
      stop(paste("Error: Data frame for condition", condition_names[i], "must contain 'log2FoldChange' and 'pAdj' columns."))
    }
    
    # Count the number of positively and negatively differentially expressed genes
    positive_genes[i] <- sum(df$log2FoldChange > logfc_threshold & df$padj < padj_threshold, na.rm = TRUE)
    negative_genes[i] <- sum(df$log2FoldChange < -logfc_threshold & df$padj < padj_threshold, na.rm = TRUE)
  }
  
  # Combine the counts into a matrix for the barplot
  gene_counts <- rbind(negative_genes, positive_genes)
  
  # Calculate the range for the y-axis
  y_min <- min(c(0, negative_genes))  # Start from 0 to include positive values
  y_max <- max(c(positive_genes, abs(negative_genes))) * 1.2  # Extend the limit a bit to leave space for the labels
  
  # Plot the stacked barplot
  bar_positions <- barplot(gene_counts,
                           beside = TRUE,  # Plot bars side by side
                           names.arg = condition_names,  # Condition labels
                           col = c("red", "green"),  # Colors for positive and negative
                           main = title,
                           ylab = "Number of Genes",
                           ylim = c(y_min, y_max))  # Adjust the y-axis limits
  
  # Add a legend
  legend("topright", legend = c("Negatively DEGs", "Positively DEGs"), fill = c("red", "green"))
  
  # Add text on top of each bar showing the count
  text(x = bar_positions[1, ], y = negative_genes, labels = negative_genes, pos = 3, col = "black")  # Negative DEGs
  text(x = bar_positions[2, ], y = positive_genes, labels = positive_genes, pos = 3, col = "black")  # Positive DEGs
}
```

```{r}

L36 = list(reList$res1F_WT36, reList$res1M_WT36, reList$res2F_WT36, reList$res2M_WT36, reList$res3F_WT36, reList$res3M_WT36)

plot_DEG_barplot(condition_dfs = L36, condition_names = c("T1F", "T1M", "T2F", "T2M", "T3F", "T3M"), title = "Total Differentially Expressed RBPs - L36")
```

```{r}
L41 = list(reList$res1F_WT41, reList$res1M_WT41, reList$res2F_WT41, reList$res2M_WT41, reList$res3F_WT41, reList$res3M_WT41)

plot_DEG_barplot(condition_dfs = L41, condition_names = c("T1F", "T1M", "T2F", "T2M", "T3F", "T3M"), title = "Total Differentially Expressed Genes - L41")
```
